crates/router/src/core/payments/routing.rs

```diff
@@ -1202,3 +1208,105 @@ pub fn make_dsl_input_for_surcharge(
+            .clone()
+            .map(|val| val.parse_value("DynamicRoutingAlgorithmRef"))
+            .transpose()
+            .change_context(errors::RoutingError::DeserializationError {
+                from: "dynamic_routing_algorithm".to_string(),
+                to: "DynamicRoutingAlgorithmRef".to_string(),
+            })
+            .attach_printable(
+                "unable to deserialize dynamic_routing_algorithm from business profile to DynamicRoutingAlgorithmRef",
+            )?
+            .unwrap_or_default();
+
+    let success_based_algo_ref = success_based_dynamic_routing_algo_ref
+        .success_based_algorithm
+        .ok_or(errors::RoutingError::GenericNotFoundError { field: "success_based_algorithm".to_string() })
+        .attach_printable(
+            "success_based_algorithm not found in dynamic_routing_algorithm from business_profile table",
+        )?;
+
+    if success_based_algo_ref.enabled_feature
+        == api_routing::SuccessBasedRoutingFeatures::DynamicConnectorSelection
+    {
+        let client = state
+            .grpc_client
+            .dynamic_routing
+            .success_rate_client
+            .as_ref()
+            .ok_or(errors::RoutingError::SuccessRateClientInitializationError)
+            .attach_printable("success_rate gRPC client not found")?;
+
+        let success_based_routing_configs = routing::helpers::fetch_success_based_routing_configs(
+            state,
+            business_profile,
+            dynamic_routing_algorithm,
+        )
+        .await
+        .change_context(errors::RoutingError::SuccessBasedRoutingConfigError)
+        .attach_printable("unable to fetch success_rate based dynamic routing configs")?;
+
+        let tenant_business_profile_id = format!(
+            "{}:{}",
+            state.tenant.redis_key_prefix,
+            business_profile.get_id().get_string_repr()
+        );
+
+        let success_based_connectors: CalSuccessRateResponse = client
+            .calculate_success_rate(
+                tenant_business_profile_id,
+                success_based_routing_configs,
+                routable_connectors,
+            )
+            .await
+            .change_context(errors::RoutingError::SuccessRateCalculationError)
+            .attach_printable(
+                "unable to calculate/fetch success rate from dynamic routing service",
+            )?;
+
+        let mut connectors = Vec::with_capacity(success_based_connectors.labels_with_score.len());
+        for label_with_score in success_based_connectors.labels_with_score {
+            let (connector, merchant_connector_id) = label_with_score.label
+                .split_once(':')
+                .ok_or(errors::RoutingError::InvalidSuccessBasedConnectorLabel(label_with_score.label.to_string()))
+                .attach_printable(
+                    "unable to split connector_name and mca_id from the first connector obtained from dynamic routing service",
```

Comment:
why does message have `first connector obtained`?

Reply:
This is refactored to something like:
`unable to split connector_name and mca_id from label obtained by the dynamic routing service`